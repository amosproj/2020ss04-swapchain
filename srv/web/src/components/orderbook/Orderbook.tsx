import React, { useState } from "react"
import crypto from "crypto-random-string"
import { Table } from "./Table"
import Modal from "./Modal"
import * as bitcoin from "bitcoinjs-lib"
import { ReactComponent as LeftArrow } from "../../icons/arrow-left.svg"

export const status = {
    active: {
        label: "active",
        color: "green",
    },
    fulfilled: {
        label: "fulfilled",
        color: "gray",
    },
    expired: {
        label: "expired",
        color: "gray",
    },
}


export interface Order {
    selected: boolean,
    address: string
    created: Date
    validUntil: Date
    status: { label: string, color: string }
    give: {
        asset: string
        value: number
    }
    exchangeRate: number,
    receive: {
        asset: string
        value: number
    }
}

export const orderFactory = (count: number): Order[] => {
    let orders: Order[] = []
    for (let i = 0; i < count; i++) {
        const value = Number((Math.random() / 1_000).toFixed(10));
        const rng = Math.random()
        const validUntil = new Date(new Date().getTime() + 86_400_000 * (Math.random() - 0.1) * 7)
        orders.push({
            selected: false,
            address: bitcoin.crypto.sha256(Buffer.from(crypto({ length: 64, type: "base64" }))).toString(),
            created: new Date(),
            validUntil,
            status: (validUntil < new Date()) ? status.expired : (rng > 0.5) ? status.active : status.fulfilled,
            give: {
                asset: (rng > 0.5) ? "BTC" : "BTS",
                value: value,
            },
            exchangeRate: rng + 0.5,
            receive: {
                asset: (rng > 0.5) ? "BTS" : "BTC",
                value: value * (rng + 0.5),
            }
        })
    }

    orders[0].address = "abcd"

    return orders
}



type Props = {
    orders: Order[]
}


export default (props: Props) => {
    const { orders } = props
    const [modalOpen, setModalOpen] = useState(false)
    const [menuOpen, setMenuOpen] = useState(true)
    const [selectedOrder, setSelectedOrder] = useState(orders[0])

    const selectOrder = (order: Order): void => {
        setSelectedOrder(order)
        setModalOpen(true)
    }


    return (


        <div className="flex">
            <nav className="p-4 bg-white border-r border-gray-400 w-max-sm">
                <div className="flex flex-col ">
                    <span className="px-4 py-2 text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase ">Status</span>
                    <div>

                        {[status.active, status.expired, status.fulfilled].map(s => {
                            return (
                                <button className="px-2 py-1 mx-2 text-xs font-semibold text-gray-800 uppercase bg-gray-100 rounded-sm hover:bg-gray-300 hover:text-gray-900">{s.label}</button>
                            )
                        })}
                    </div>
                </div>
                <div className="mt-8">
                    <span className="px-4 py-2 text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase ">Address</span>

                    <div className="relative ml-4">
                        <span className="absolute inset-y-0 left-0 flex items-center pl-3">
                            <svg className="w-6 h-6 text-gray-600" fill="none" viewBox="0 0 24 24" stroke="currentColor">
                                <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z" />
                            </svg>
                        </span>
                        <input className="block py-3 pl-10 text-gray-900 placeholder-gray-600 border border-gray-400 rounded-md" placeholder="Address"></input>
                    </div>
                </div>

            </nav>
            <div className="flex-grow">

                <Table orders={orders} onRowClick={selectOrder} ></Table>

            </div>

            <Modal order={selectedOrder} open={modalOpen} close={() => setModalOpen(false)}></Modal>
        </div>
    )
}