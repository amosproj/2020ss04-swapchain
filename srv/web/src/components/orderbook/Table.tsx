import React, { useState } from "react"
import { Order } from "./Orderbook"
import { Row } from "./Row"
import { ReactComponent as ChevronDown } from "../../icons/chevron-down.svg"
import { ReactComponent as ChevronUp } from "../../icons/chevron-up.svg"
import { ReactComponent as Exclamation } from "../../icons/exclamation.svg"

type Props = {
    orders: Order[],
    onRowClick: (order: Order) => void
}


export const Table = (props: Props) => {
    const { orders, onRowClick } = props


    const [sortedByDate, setSortedByDate] = useState(0)
    const [sortedByRate, setSortedByRate] = useState(0)
    const [sortedByAmountYouPay, setSortedByAmountYouPay] = useState(0)


    const toggleSortByDate = () => {
        switch (sortedByDate) {
            case -1:
                orders.reverse()
                setSortedByDate(1)
                break;
            case 0:
                orders.sort((a: Order, b: Order) => { return b.created.getTime() - a.created.getTime() })
                setSortedByDate(1)
                break;
            case 1:
                orders.reverse()
                setSortedByDate(-1)
                break;

        }
    }
    const toggleSortByRate = () => {
        switch (sortedByRate) {
            case -1:
                orders.reverse()
                setSortedByRate(1)
                break;
            case 0:
                orders.sort((a: Order, b: Order) => { return b.exchangeRate - a.exchangeRate })
                setSortedByRate(1)
                break;
            case 1:
                orders.reverse()
                setSortedByRate(-1)
                break;

        }
    }
    const toggleSortByAmountYouPay = () => {
        switch (sortedByAmountYouPay) {
            case -1:
                orders.reverse()
                setSortedByAmountYouPay(1)
                break;
            case 0:
                orders.sort((a: Order, b: Order) => { return b.give.value - a.give.value })
                setSortedByAmountYouPay(1)
                break;
            case 1:
                orders.reverse()
                setSortedByAmountYouPay(-1)
                break;

        }
    }





    return (
        <div className="flex-grow min-w-full bg-white ">
            <div className="flex items-center justify-between">
                <div className="flex items-center">
                    <h2 className="p-6 text-2xl text-gray-700">Orderbook</h2>
                    <span className="px-2 text-sm text-gray-700 bg-gray-200 rounded">{orders.length}</span>
                </div>
                <div className={` flex items-center px-4 py-2 mr-8 bg-yellow-100 rounded transition duration-300 transform ease-in-out ${orders.length > 0 ? "opacity-0 translate-x-20" : "opacity-100"}`}>
                    <Exclamation className="w-4 h-4 text-orange-700"></Exclamation>
                    <span className="ml-3 text-xs font-bold text-orange-900">
                        Looks like there is nothing here, try removing some filters.
                        </span>
                </div>
            </div>
            <div className="overflow-x-auto">
                <div className="inline-block min-w-full overflow-hidden">
                    <table className="min-w-full table-auto">
                        <thead className="border-b border-gray-300">
                            <tr>
                                < th className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100">
                                    <div className="flex">
                                        <span>Created</span>
                                        {sortedByDate === 1 ?
                                            <ChevronUp onClick={toggleSortByDate} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronUp>
                                            :
                                            <ChevronDown onClick={toggleSortByDate} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronDown>
                                        }
                                    </div>
                                </th>
                                <th className="flex px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100">
                                    Status
                        </th>
                                <th className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100">
                                    <div className="flex">
                                        <span>Amount You Pay</span>
                                        {sortedByRate === 1 ?
                                            <ChevronUp onClick={toggleSortByAmountYouPay} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronUp>
                                            :
                                            <ChevronDown onClick={toggleSortByAmountYouPay} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronDown>
                                        }
                                    </div>
                                </th>
                                <th className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100">
                                    <div className="flex">
                                        <span>Rate</span>
                                        {sortedByRate === 1 ?
                                            <ChevronUp onClick={toggleSortByRate} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronUp>
                                            :
                                            <ChevronDown onClick={toggleSortByRate} className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"></ChevronDown>
                                        }
                                    </div>
                                </th>
                            </tr>
                        </thead>
                        <tbody className="bg-white">
                            {orders.map((order, key) => {
                                return (
                                    <Row key={key} order={order} onClick={onRowClick}></Row>
                                )
                            })}
                        </tbody>
                    </table>

                </div>
            </div >
        </div>
    )
}