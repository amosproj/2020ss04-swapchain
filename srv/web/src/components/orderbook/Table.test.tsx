import React from 'react';
import {Order} from './Orderbook';
import {Table} from "./Table"
import renderer from 'react-test-renderer';

jest.spyOn(Math, "random").mockImplementation(() => 1)
jest.spyOn(Date.prototype, "toLocaleString").mockImplementation(() => "IAmADate")
jest.spyOn(Date.prototype, "toLocaleTimeString").mockImplementation(() => "IAmADate")


it('renders correctly', () => {
    const order: Order = {
        selected: true,
        addressHash: "asdf",
        created: new Date(),
        validUntil: new Date(),
        status: { label: "label", color: "red" },
        give: {
            asset: "BTC",
            value: 1
        },
        exchangeRate: 1,
        receive: {
            asset: "BTS",
            value: 1,
        }
    }

    const orders = []
    for (let i = 0; i < 10; i++) {
        orders.push(order)
    }

    const tree = renderer
        .create(<Table orders={orders} onRowClick={()=>{}}></Table>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});