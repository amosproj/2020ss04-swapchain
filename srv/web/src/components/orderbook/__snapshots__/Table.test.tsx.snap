// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex-grow min-w-full bg-white "
>
  <div
    className="flex items-center justify-between"
  >
    <div
      className="flex items-center"
    >
      <h2
        className="p-6 text-2xl text-gray-700"
      >
        Orderbook
      </h2>
      <span
        className="px-2 text-sm text-gray-700 bg-gray-200 rounded"
      >
        10
      </span>
    </div>
    <div
      className=" flex items-center px-4 py-2 mr-8 bg-yellow-100 rounded transition duration-300 transform ease-in-out opacity-0 translate-x-20"
    >
      <svg
        className="w-4 h-4 text-orange-700"
      >
        exclamation.svg
      </svg>
      <span
        className="ml-3 text-xs font-bold text-orange-900"
      >
        Looks like there is nothing here, try removing some filters.
      </span>
    </div>
  </div>
  <div
    className="overflow-x-auto"
  >
    <div
      className="inline-block min-w-full overflow-hidden"
    >
      <table
        className="min-w-full table-auto"
      >
        <thead
          className="border-b border-gray-300"
        >
          <tr>
            <th
              className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100"
            >
              <div
                className="flex"
              >
                <span>
                  Created
                </span>
                <svg
                  className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"
                  onClick={[Function]}
                >
                  chevron-down.svg
                </svg>
              </div>
            </th>
            <th
              className="flex px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100"
            >
              Status
            </th>
            <th
              className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100"
            >
              <div
                className="flex"
              >
                <span>
                  Amount You Pay
                </span>
                <svg
                  className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"
                  onClick={[Function]}
                >
                  chevron-down.svg
                </svg>
              </div>
            </th>
            <th
              className="px-6 py-3 text-xs font-medium leading-4 tracking-wider text-left text-gray-700 uppercase bg-white-100"
            >
              <div
                className="flex"
              >
                <span>
                  Rate
                </span>
                <svg
                  className="w-4 h-4 ml-1 text-gray-600 hover:text-gray-900"
                  onClick={[Function]}
                >
                  chevron-down.svg
                </svg>
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          className="bg-white"
        >
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
          <tr
            className="border-l-2 border-transparent hover:bg-gray-100 hover:border-gray-400 bg-gray-100 border-l-4 border-gray-400"
            onClick={[Function]}
          >
            <td
              className="table-cell px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex flex-col text-sm leading-5"
              >
                <span
                  className="text-gray-700"
                >
                  IAmADate
                </span>
                <span
                  className="text-xs text-gray-500"
                >
                  IAmADate
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 text-left whitespace-no-wrap border-t border-gray-200"
            >
              <span
                className="px-2 text-xs uppercase font-semibold rounded bg-red-100 text-red-800"
              >
                label
              </span>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="inline-flex px-2 text-xs font-semibold uppercase bg-gray-300 rounded "
                >
                  BTC
                </span>
                <span
                  className="ml-2 font-mono leading-5 text-gray-500"
                >
                  1
                </span>
              </div>
            </td>
            <td
              className="px-6 py-4 whitespace-no-wrap border-t border-gray-200"
            >
              <div
                className="flex items-center"
              >
                <span
                  className="flex items-center ml-2 leading-5 text-gray-500"
                >
                  1.00000
                  <div
                    className="flex flex-col ml-4 text-xs tracking-tight text-gray-500"
                  >
                    <span>
                      BTS
                    </span>
                    <div
                      className="border-t border-gray-400 "
                    />
                    <span>
                      BTC
                    </span>
                  </div>
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
